### DISCUSSION TITLE
So generic

### DISCUSSION THREAD

**Discussion by dodexahedron** [dodexahedron/TerminalGuiDesigner#32] [Ideas]
Ok, so I figured this is a better place to talk about the work on generics than scattered amongst issues and PRs...

Anyway...

Let's talk about it!

**Comment by dodexahedron** [comment]
In code I have not yet committed or pushed, I went ahead and took a stab at seeing just how far I would have to take generics, in a minimal implementation of them, without making ALL code generic.

Ended up having to touch most classes, anyway, and the mixture of non-generic and generic code leaves some...let's say "fun" artifacts, here and there, but it compiles lol. It was over 300 individual changes, and many of them aren't how I'd like them to be, since I often had to fall back to `View` as a type parameter, in the interest of time. That's fine, since everything inherits from `View`, but the end goal would be to have actual type parameters nearly everywhere, so things are as tight as possible.

Problem is, about 1/4 of the tests fail (there are now over 1000 tests), and it seems to be due to references not being properly maintained by the AddViewOperation class, after compiling code generated by ViewToCode, and reversing it all. I'm trying to debug the cause of that, because it appears to _also_ be the same root problem I was having in another branch, without the more extensive generic changes (just the ViewFactory, in that branch). I am not yet sure if it is the unit test project and technicalities of how it executes causing it, or if it's a subtle problem that the new genericized type system created or exposed in CodeToView (where everything falls apart). I haven't yet tried to actually _run_ it, so I suppose I should do that, to see if the tests are simply no longer valid, with the new code. I'll do that after I finish rambling.

The generated code still looks great, but the compiled code doesn't get interpreted properly, and, for some reason, the subview (which is in code and looks correct to me) doesn't make it back.

Here's an example designer.cs created by a test method with that code:

```c#

//------------------------------------------------------------------------------

//  <auto-generated>
//      This code was generated by:
//        TerminalGuiDesigner v1.1.0.0
//      Changes to this file may cause incorrect behavior and will be lost if
//      the code is regenerated.
//  </auto-generated>
// -----------------------------------------------------------------------------
namespace YourNamespace {
    using System;
    using Terminal.Gui;
    
    
    public partial class Test60Percent_RoundTrip : Terminal.Gui.Window {
        
        private Terminal.Gui.Label myViewOut;
        
        private void InitializeComponent() {
            this.myViewOut = new Terminal.Gui.Label();
            this.Width = 10;
            this.Height = 5;
            this.X = null;
            this.Y = null;
            this.Visible = true;
            this.Modal = false;
            this.TextAlignment = Terminal.Gui.TextAlignment.Left;
            this.Title = "";
            this.myViewOut.Width = Dim.Percent(60f);
            this.myViewOut.Height = 1;
            this.myViewOut.X = Pos.Percent(60f);
            this.myViewOut.Y = 0;
            this.myViewOut.Visible = true;
            this.myViewOut.Data = "myViewOut";
            this.myViewOut.Text = "Heya";
            this.myViewOut.TextAlignment = Terminal.Gui.TextAlignment.Left;
            this.Add(this.myViewOut);
        }
    }
}
```
During execution of that `Tests.RoundTrip` method, in the test that generated the above code, the Label (myViewOut), for some reason, isn't a subview of the Window (in fact, Subviews on that Window is empty), after the call to `CodeToView.CreateInstance` returns a new Design object.

I have a bunch of ideas of what could be happening, here, but I have no definitive answer, yet.\
An incomplete list of things I'm considering, right now, are:
 - Issues with statics and reference equality on returned constructed instances
 - Issues related to using generics, without other modification to code that may now be more strict, due to slightly different type rules at run-time (this one seems likely to be at least part of it all)
   - Also, this could be exacerbated by having a mixture of generic and non-generic code, but I have no concrete (yay puns) examples of that being definitively true, yet
 - There's still a fair amount of reflection going on, especially around CodeToView, so I'm wondering if maybe the reflected types aren't being treated as equivalent to their original types, at run-time.
   - Most of that reflection shouldn't be necessary, in theory, since we are already loading the assembly, and thus _should_ be able to treat types it contains as if they are part of the appdomain (meaning generics should work implicitly)
 - Something subtle happening due to not using generics _everywhere_ (hmm...a theme here?)

And a few other things that just aren't coming to mind while I'm trying to list them right now....

First thing I'm going to do, right after I finish this comment, is go build it on my branch with the more extensive generic usage, and see if it works properly (or at all 😅), which will at least tell me if I need to re-focus on just re-working the tests, instead of continuing investigation in the core of it all.

Another thing I'll do, likely in either case, is bite the bullet and genericize the Design class, which is one of the few big ones that didn't get the generic treatment, yet.

**Reply by tznind** [reply]
> There's still a fair amount of reflection going on, especially around CodeToView, so I'm wondering if maybe the reflected types aren't being treated as equivalent to their original types, at run-time. 

This may be the case, especially for the root view that the user is designing.

Calling `Open` will create a new assembly in memory by compiling the `.Designer.cs` file.

In this case the `MyWindow` created as a result of `Open` will be a different Type (in a different 'assembly') than the original.

This is because Designer.cs processing is done by compilation not parsing.  So every time you open a Designer.cs you get a new assembly with a new Type reference for `MyWindow`.

When you create with `New` I think the process is more or less the same, behind the scenes it is just creating the 'bare bones' and then calling 'Open' on the file generated.

**Reply by dodexahedron** [reply]
Yeah, I think it's basically just breaking the inheritance chain, when we create them with the Activator, because the full type name would include the assembly it came from, which has a GUID name. Thus, when I try to use the generic `ViewFactory.Create<T>()`, the concrete type generated at run-time probably isn't implicitly assignable to and from the newly-compiled type. They both still inherit from Terminal.Gui.View, so they can be downcast to that, but that's of course not a real solution anyway, especially when it happens to the root view, since that's the only one that's really a unique type, even if it inherits from View, Window, Dialog, etc., in the various test cases.

I just tried replacing that usage of the Activator in `CodeToView.CreateInstance()` with a call to `ViewFactory.Create()`, a second ago, and all but 2 tests continue to pass, with no other modifications, which is encouraging. The ones that are failing are in DragOperationTests: `TestSimpleDrag_IntoTabView()` and `TestSimpleDrag_OutOfFrameView_IntoRootWindow`. The other tests in that fixture pass, so I'm betting it's just something simple in the test itself. Looks to be hit-test related, based on the initial exception I'm seeing before investigating.

Since the assembly is being loaded into the AppDomain, we can treat the types it contains as normal run-time types not needing reflection, which means generics will ultimately work. It's just going to take a more measured and gradual approach to get there. That little expedition I went on to just try using the generic Create there, and then modifying broken calls and types to generics, where errors existed, was the low-ish effort, naive attempt, with fingers crossed that it would be enough to make it all "just work," but, unfortunately, I just wasn't that lucky.

I imagine that, as reflection is replaced, even with the non-generic methods that pass types around, that situation will get closer and closer to being doable with a lot more certainty. After all, once we know, at all points in the code, that a given instance of a `Type` is valid, that is a future point at which a generic will be useable, so long as it doesn't have to depend on a reflected type from a different-named assembly throughout its lifetime.

Creating types via reflection is _almost_ the same as doing it the normal way, but you do get that baggage with type name mangling that presents subtle issues like this, if you try to mix that approach with generics.

And, as mentioned in one of my follow-up comments to myself, below, we'll eventually gain support for some of Terminal.Gui's generic types more or less "for free," once we are able to actually take that plunge. We could special-case it, now, with reflection, by checking if the type is generic and then going down a different path that builds the appropriate generic typeinfos, but that's work that would be wasted to do now, rather than just waiting til we have native generic support throughout, since that's already a known limitation in existing versions.

Even with what little of that I've done, in the branch I'm now working on, it already makes some tests which previously failed on known unsupported types pass, if taken out of the unsupported collections, which indicates we're inching closer already. 🙂

I'm planning on converting some more tests to NUnit 4 and the Constraint model, tonight, so you should probably see some new commits on the PR, later tonight (or this morning, in your case!). Already have a couple of commits that will be part of that. 🥳

**Reply by tznind** [reply]
Great! I would like to merge https://github.com/gui-cs/TerminalGuiDesigner/pull/260 when it is ready.  The tests all pass and I am liking the changes very much.

I see the other branches you are working on but am not sure which are investigations and which are the next/main bit of work to merge.  Once you have a direction to go in for this generics can you please open a draft PR (after the unit test one is merged so the diffs will be clearer)?

We have to be careful with `T : View, new( )` because one of the goals down the line is to support adding user's own custom classes e.g.

- I create Inventory.Designer.cs
- I create Room.Designer.cs
- I want to create MyGame.Designer.cs and add both my custom views to this

At this point we would be forcing the user to have a blank constructor.  It would prevent us doing something like e.g. calling the constructor with null/default values or doing some other kind of logic.

This would especially be a problem if the `new ()` constraint on `T` became viral and appeared on classes other than `ViewFactory` e.g. `AddViewOperation`

**Reply by dodexahedron** [reply]
The main work is the one on that PR. That's the main one to follow. Others get merged into that when they're complete.

I'm only working on unit tests, right now, and likely won't get back to the other stuff until I'm done or close to it.

Just finished with CopyPasteTests a few minutes ago and pushed, so you can now see how those are coming along. Those tests got some pretty significant upgrades.

Next will probably be finishing up work on the DimTests fixture, and then on from there.

**Reply by dodexahedron** [reply]
So, thinking about your concern about the `new()` constraint a bit more:

TGD already had an implicit restriction on views having a public parameterless constructor available, for anything that called ViewFactory.Create that wasn't caught by any of the specific type filters, because the reflection at the end of it called the parameterless constructor. Therefore, the new() constraint is a non-breaking change for anything calling Create.

In a way, it's actually a "fix," since now there is a compile-time guarantee of type safety whereas, before, if there was no public parameterless constructor for a type instantiated by the Create method, it would have been a runtime error thrown by the Activator.

For Create to support other constructors, it would need to accept one or two collection inputs - one in any case, for parameters to pass to constructors and for use in selecting them, or two if supporting generic constructors were to be desired, to give the type parameters.

Now... For as long as the non-generic Create method exists, runtime exceptions can still occur, as they would have before, for code calling the non-generic version, since static analysis can't predict that. But that's also identical behavior to the previous code. However, that case does now also have an `IsAssignableTo(typeof(View))` check, to at least throw a predictable exception at run-time, if any such improper uses do exist (there aren't any, in TGD itself). I'll likely add unit tests to formally cover that, when I flesh out the ViewFactory tests a bit more later on.

In the end, the constraint on the generic method is a non-change and the behavior change in the non-generic method is _almost_ a non-change, since it'll still fail or succeed in the same situations, just with an `ArgumentOutOfRangeException` on t, when t is not a descendent type of `View`.

**Reply by tznind** [reply]
> In a way, it's actually a "fix," since now there is a compile-time guarantee of type safety whereas, before, if there was no public parameterless constructor for a type instantiated by the Create method, it would have been a runtime error thrown by the Activator.

This is great work, I am fully committed to the new generics approach.  But I would like to avoid adding `new ()` as a Type constraint.

I realize that it is currently a requirement of the existing `ViewFactory`.  But in future it may not be.  It will be much easier to add support for parameter constructors if there is only one place we need to change (i.e. `ViewFactory`).  If we have to back out `new ()` from Type constraints system wide - it becomes a much bigger task.

I left a comment on the PR that a way of avoiding `new ()` to start with would be using:

```
T newView = (T)ReflectionHelpers.GetDefaultViewInstance(typeof(T));
```

### Rationale

Currently all Views in terminal gui have parameterless constructors - and its one of the design requirements for new Views (❤️ @tig)

In TGD, the view being designed (MyView.Designer.cs) is currently **compiled**.  As follows: 
- The MyView.Designer.cs file on disk
- Plus a system generated (memory only) MyView.cs (i.e. not the one the user has written) that contains only blank constructor and `InitializeComponent` call
- Type References from the core dotnet framework reference assemblies and Terminal.Gui library

Since we generate the constructor code ourselves in memory - we are in charge and can always make it parameterless.  Even if the user declares parameterized constructors in their own code it won't affect anything.  It also means that when designing they don't see any effects of code they add to the constructor which is... probably for the best.

But I'm not sure how maintainable that approach is when it comes to custom Views (e.g. I want to design MyView.Designer.cs and add MyOtherView.Designer.cs as a child View).

There are really 2 approaches to getting instances of the users Views
- Limitted Compilation of users code (as described above)
- Dll hoovering (take custom `View` types from the compiled dlls of the users library).  I think this is what WinForms designer did - and it has its own issues.

We may need to mix and match approaches down the line

At the moment I would really just like to avoid us being hard wed to parameterless constructors throughout the whole tool at a generics level.

**Reply by dodexahedron** [reply]
I gotta get to bed, but I just saw this in the thread, so I'll look it over tomorrow, with a fresh head. :)

Thanks for the analysis.

**Reply by dodexahedron** [reply]
Okie dokie, now I have some time to sit down and give your thoughts the attention they deserve. :)

> I realize that it is currently a requirement of the existing ViewFactory. But in future it may not be. It will be much easier to add support for parameter constructors if there is only one place we need to change (i.e. ViewFactory). If we have to back out new () from Type constraints system wide - it becomes a much bigger task.

Certainly a valid and laudable goal. And I think some stuff you say further down actually paves a better way forward, for TGD, which I'll address there.

My concern with that proposal, as-is, is just that it goes the _opposite_ direction, and makes _all_ usages of Create<T> reflection anyway, which kinda negates the point of it in the first place. I liked how the old method did it, where it tried known supported types, and then used reflection as a fallback, and it was my intent for the generic one to do the same - I just messed up and didn't get all the code into this branch that I had written on the other branch.

As general commentary on ViewFactory, itself:\
ViewFactory is pretty much just a convenience shortcut to calling constructors, whether in-app or via reflection. But, the real meat of TGD is the Editor and Design classes. In theory, the ViewFactory isn't really even something that user code should ever call, directly, precisely because it is an internal method proxy helper class, for the most part. User code shouldn't have any dependency on the internals of TGD, since they're targeting TG.

Anyway, back to your thoughts:

> Currently all Views in terminal gui have parameterless constructors - and its one of the design requirements for new Views (❤️ @tig)

Yeah. And, in general, I do tend to agree with that design decision, specifically for this kind of application, especially since the underlying library (TG) already has that restriction. _BUT!_ Your next ideas sound good to me and get me thinking...

> Since we generate the constructor code ourselves in memory - we are in charge and can always make it parameterless. Even if the user declares parameterized constructors in their own code it won't affect anything. It also means that when designing they don't see any effects of code they add to the constructor which is... probably for the best.

Good thinking. 👍 \
This I think is potentially the goldilocks solution, even if it may require a little extra code on our part. We can load the user code as-is (before compiling), check the syntax tree for a parameterless constructor, and add an internal one, if necessary, before doing the compile. Roslyn makes that not too difficult, and it would be seamless to the user. We can just keep a private/internal flag around to inform the ViewToCode whether to skip emitting that constructor, when generating code for output.

It would enable your goal of allowing users the extra flexibility in their custom views of being allowed to have parameterized constructors (maybe even generic, with a bit more work on our end), while also basically providing the "glue" between their code and TG, so that we don't have to special case anything else and can enjoy compile-time type safety both at TGD compilation and at user code dynamic compilation. I like the thought and totally think we should explore it (maybe once TG v2 work is closer to completion?).

As for this:

>But I'm not sure how maintainable that approach is when it comes to custom Views (e.g. I want to design MyView.Designer.cs and add MyOtherView.Designer.cs as a child View).

Hmm. Good thought, but I'm wondering just how far we should try to take it. A fully recursive capability that doesn't enforce at least some restrictions can quickly balloon out of control, and such a use case is probably (just my assumption) a minority of users, so I'd just want us to be aware of cost/benefit. The restriction that derived classes need to have an available (even if internal or protected) parameterless constructor isn't really an onerous one. And we aren't requiring them to actually _use_ that constructor, anyway. Currently, it simply has to exist, either explicitly or implicitly.

Actually, that gives me an idea... What if we were to simply wrap user code in a class that directly inherits from View or an appropriate descendent class, for purposes of compilation at design time (well...not really wrapping....it's inheritance)? That would fit in with your above idea as well as give us and the users extra freedom, quite possibly without imposing direct restrictions on their code. Obviously we would only output _their_ class, leaving out our intermediate ancestor of it. It would also enable us to do potentially useful things in TGD, down the line, like injecting code into that class that is used at design time but not needed for output. Maybe something else worth exploring. And then the whole constructor thing becomes totally moot, because TGD simply wouldn't have to care what constructors the user has written or removed. It also could potentially enable users to use the TGD unit test suite against their own code, which could be a pretty cool win for them.

> At the moment I would really just like to avoid us being hard wed to parameterless constructors throughout the whole tool at a generics level.

This discussion I think is leading to a place where we might be able to get the best of both worlds, if we give it some good thought.

Good stuff!

Let's keep thinking about it.

In the meantime, I'll be sure to formalize some tests of ViewFactory to ensure 1:1 equivalence with the old code, at least, so that we can keep moving on and explore all these ideas in their own context. :)

**Reply by dodexahedron** [reply]
>Actually, that gives me an idea... What if we were to simply wrap user code in a class that directly inherits from View or an appropriate descendent class, for purposes of compilation at design time (well...not really wrapping....it's inheritance)? That would fit in with your above idea as well as give us and the users extra freedom, quite possibly without imposing direct restrictions on their code. Obviously we would only output their class, leaving out our intermediate ancestor of it. It would also enable us to do potentially useful things in TGD, down the line, like injecting code into that class that is used at design time but not needed for output. Maybe something else worth exploring. And then the whole constructor thing becomes totally moot, because TGD simply wouldn't have to care what constructors the user has written or removed. It also could potentially enable users to use the TGD unit test suite against their own code, which could be a pretty cool win for them.

Another thought on this....

That actually could potentially be a natural evolution of the Design class.
Design could be genericized, as one way of achieving this. I mean, Design probably should be genericized anyway, for other reasons that would have nice benefits, but it does also give one approach to these ideas a place to live without having to create something entirely new.

**Reply by dodexahedron** [reply]
Would love to do a whiteboard session or something at some point in the future so we can brainstorm in real time, if you're up for it.

Probably not at least til I'm done with the unit test migration, though, so there aren't too many balls in the air. We're only 2 people volunteering time, after all. 😅

**Reply by tznind** [reply]
Ok yeah you have convinced me, lets go with `T: View, new ( )` 

Thinking about it, we currently only look at the `.Designer.cs` file anyway when it comes to compiling.  So we should be able to follow the same approach to user views.

```
// foreach member in .Designer.cs 
   // if we do not know this Type
      // locate file
           // Add file's .Designer.cs to 'compilation source code'
          // Create virtual .cs file with blank constructor (calling `InitializeComponent`)
```

**Reply by dodexahedron** [reply]
Apparently email reply doesn't keep it in the thread. Re-posting this in-thread:

Haha well that wasn't exactly the goal, and we likely won't need it too many places anyway. And using your idea plus maybe some of the other stuff mentioned above, we can likely never even have to expose that to the user-facing API.

**Reply by dodexahedron** [reply]
Okie dokie. So...

I have written a pretty thorough test (but with some intentional warnings) that goes over every supported view class, makes one via each Create method, and then compares all of their public properties via reflection, to ensure the two returned Views are, in fact, the same.

What do I mean by intentional warnings? Well, Terminal.Gui doesn't define equality operators for most classes, so I'm going to have to write some comparisons that fit the needs of this test, for purposes of checking certain properties on a handful of types. The values look the same to me (via debugger inspection), but Is.EqualTo doesn't work on them, since they aren't IEquatable, so that ends up just being a reference equality check (which of course fails, as one would expect).

Putting them up as passing with warnings was to allow you to get a look at the tests and the slightly revised ViewFactory code that the tests revealed were either necessary or a good idea (for both). If you have any ideas for additional things to check or any shortcuts that might be possible for checking the properties with warnings on them, I'm all ears! 😅  Not really looking forward to that, though I suppose that work could be shared with Terminal.Gui itself, depending on final implementation.

I had considered using the code generation already in TGD to just write them out and compare code, but that's a circular test dependency, so bummer there haha. Oh well. I'll probably work on finishing this up tomorrow.

Latest state of the branch: \
https://github.com/dodexahedron/TerminalGuiDesigner/tree/convert-unit-tests-to-constraint-model-2

Here's a shortcut link to the first line of the bigger, more important test method I just wrote: \
https://github.com/dodexahedron/TerminalGuiDesigner/blob/convert-unit-tests-to-constraint-model-2/tests/ViewFactoryTests.cs#L32

**Reply by dodexahedron** [reply]
Oh before I forget:

One of the improvements gained from having this test (whether generics are involved or not) is that it sussed out some types that were falling back to reflection and then having unnecessary properties set on them. In particular, those are `Line`, `TileView`, and `Slider`. Since the reflection fallback always calls SetActualText, that was happening on them, setting the Text property. If any of those _should_ be getting that property set, let me know and I'll move them to the block above them, which does set that property.

**Reply by tig** [reply]
Should we make View derive from IEquatable?

**Reply by tznind** [reply]
> Should we make View derive from IEquatable?

My gut feeling is probably no.  How often does an API user want to create 2 different view instances and compare member equality (including position, text etc).  Think this is quite a niche use case.

**Reply by tznind** [reply]
> Here's a shortcut link to the first line of the bigger, more important test method I just wrote: 

Nice, that is a beast of a reflection but looks very comprehensive <3 

> Line, TileView, and Slider [...] if any of those should be getting that property set, let me know and I'll move them to the block above them, which does set that property.

That makes sense, nice catch.  No point setting `Text` on things that don't actually render it anyway.

As an asside though, TGD doesn't yet support `TileView` or `Slider`

- TileView needs to have 'extra operations' defined to be useful (e.g. split horizontal/split vertical)
- Slider uses generics so we need a way to handle that (might get a lot easier after I handle https://github.com/gui-cs/TerminalGuiDesigner/issues/267)

**Reply by dodexahedron** [reply]
Ah OK. I'll filter those out in the supportedtypes, too. They pass the filter due to there being non-generic versions of them.

Or I can leave them in, since, if we don't expose them in the UI, they're non-impacting and already there for if/when we _can_ offer support for them.

What do you think? Leave em? Delete em? Filter them in the supported linq?

**Reply by dodexahedron** [reply]
> Should we make View derive from IEquatable?

@tig 

I'm of mixed mind about that. On the one hand, it could be useful for this and other API usages, but, as @tznind noted, that's likely a more rare use case, outside of TGD.

But, at least defining the methods that interface requires on View, without actually inheriting the interface, would enable the use without burdening descendants of View with being required to implement them.

Maybe that would be a more ideal midpoint? Equality of descendants can then be implemented as desired  (such as for a ComboBox, which mostly probably needs its list inspected), and delegate to base.Equals, etc, otherwise. In fact, those types of views are the ones I'm actually most interested in, for testing purposes.

**Reply by dodexahedron** [reply]
@tznind 

For the rest of what I need to do for that beast of a test, it seems to me the remainder shouldn't be so difficult. I've already got cases in that switch to specifically filter them out and warn they aren't yet tested, so it shouldn't be much work to compare what comes back from those properties to the same expected items, by their Text and whatnot.

Initially I was thinking about it from a broader scope of enabling users to reuse the tests on their types, which was making me dread likely having to do something recursively. But that's a bit too ambitious I think. 😅

**Reply by dodexahedron** [reply]
@tig 

I think what would be probably one of the biggest helps to various things TGD does is to make a lot of private entities internal and put TerminalGuiDesigner in the InternalsVisibleTo.

Especially things like the sub-classes of Pos and Dim, for example.

Although I'm kinda of the mind that those particular types should be public anyway.

**Reply by tznind** [reply]
> What do you think? Leave em? Delete em? Filter them in the supported linq?

Lets leave them in since its tests only.  The UI is what matters and they don't show in that.  Extra test coverage is never bad to have.

>  [...] put TerminalGuiDesigner in the InternalsVisibleTo.
> Especially things like the sub-classes of Pos and Dim, for example.

This would definetly make some things in TGD less hacky and add compile time safety <3 I didn't know that it was possible to do InternalsVisibleTo across repos (guess I hadn't really thought about it).

**Reply by tig** [reply]
> > What do you think? Leave em? Delete em? Filter them in the supported linq?
> 
> 
> 
> Lets leave them in since its tests only.  The UI is what matters and they don't show in that.  Extra test coverage is never bad to have.
> 
> 
> 
> >  [...] put TerminalGuiDesigner in the InternalsVisibleTo.
> 
> > Especially things like the sub-classes of Pos and Dim, for example.
> 
> 
> 
> This would definetly make some things in TGD less hacky and add compile time safety <3 I didn't know that it was possible to do InternalsVisibleTo across repos (guess I hadn't really thought about it).

We should be very intentional about anything we change from private to internal. IMO 'internal == public' from a quality perspective. Today we are lacking good primitive unit tests for tons of public apis, and for internal it's even worse. 

Likewise for clear, complete, and concise API docs. 

A side benefit is internal actually ENABLES unit tests.

So, I'm a fan as long as we do the unit test work along the way. And we are as surgical as possible.

**Reply by dodexahedron** [reply]
>A side benefit is internal actually ENABLES unit tests.

From my current perspective (writing tests), it's the _primary_ benefit. 😅

**Reply by dodexahedron** [reply]
> I didn't know that it was possible to do InternalsVisibleTo across repos

Yup. It's just an assembly attribute pointing at a namespace name. :)

**Comment by dodexahedron** [comment]
Oh!

An upshot of some of this work, beyond the obvious benefits we gain from the generics, is that some of the previously "unsupported" classes actually can pass tests they used to be expected to fail, before. So, once I get the root issue of the round tripping figured out, that might be a really nice bonus we get out of all of this, "for free," as well. 🙂

**Comment by dodexahedron** [comment]
Ok...\
So...

I have it down to 1/5 of the tests failing, now.... So that's neat....

But, more importantly....

I believe I am narrowing down the source of _one_ of the problems I'm running into.

It's necessary to include `new()` in the type constraint on any method that has to construct a new instance of that type parameter. Thus, the ViewFactory has to have that constraint on `Create<T>()`. The resulting constraint is `where T : View, new()`. Anywhere that is called, if a specific type cannot be used in the type parameter, such as in a lot of the `*Operation` classes and anything that uses them, then the same type constraints have to be placed on those methods or their containing classes.

That's a problem, because there are types that those classes support that fail at least one of the constraints. Either there's no public parameterless constructor in the Terminal.Gui type being used (such as for operations involving MenuItems, for example), or the type does not inherit from View (such as ColorScheme).

I believe this to be at the root of a lot of the newly-failing tests, because many of them involve something that would fail one or both of these constraints, at some point in their lifecycles.

Now.... One can't just remove the type constraint, because then it's not valid to call new T(), which breaks...well...a lot of things.

And obviously we have to treat Terminal.Gui itself as a black box, so depending on any sort of change there isn't good form (though a common interface that anything designable inherits would solve a LOT of problems...).

But...
Those issues don't necessarily explain why even a label isn't being added to a window, when it should be. Or maybe it does, and it's all related to ColorSchemes, once I dig deep enough, since they don't inherit from View. And with MenuItems, there's no constructor, because those don't have a parameterless constructor.\

In any case, both of those situations can be remedied by having specific handlers for those types.

But then that starts to get in the way of working on something I mentioned elsewhere: Just using events to handle all that stuff, instead of the Operation classes. Or, even with events, those classes or something like them may still be involved, but they'd be simpler and we'd have better context with them, as well...

But that spiders out of control for what I intended to do on this branch, at least initially.

So, I think this is what I'm going to do:\
I'll make a new branch, and then commit a version of the code I'm looking at right now that compiles, but with all the problems I laid out above. I'd like to get back to the initial project, which was converting unit tests. Once that's done or at least further along, then perhaps it will be a better time to get back to this.

What started it all? Creation of that new, genericized Create method in the ViewFactory. While that, alone, isn't a problem and it creates objects of the requested type just fine, I had tried to replace uses of the old one in TerminalGuiDesigner with the generic one, and that's what created the cascade of requirements to genericize everything. Generics do tend to be somewhat viral, in a project like this. I think I will go back to an earlier commit before I embarked on that journey and focus more on just getting the tests converted, for now.

While generics are a laudable goal, there's no rush to get there.

So long as you're cool with a transition period where there are two Create methods, with identical functionality, since one just proxies to the other, for consistency. I'll cherry pick that commit on the PR branch, so you can see what that's like, @tznind .

**Comment by dodexahedron** [comment]
Copying my post from https://github.com/gui-cs/TerminalGuiDesigner/pull/264#issuecomment-1865330794 :

A commit including the re-factor of the non-generic ViewFactory.Create method is going to be included in a merge to this branch, pretty soon. It condenses the whole thing down to a single 22-line return with a switch expression that delegates each type to the generic method, using the same type-matching rules and precedence as before, so that it is a non-breaking change.

There is no longer any duplicated code between the two methods.

I also put an obsolete attribute on it, just for ease of helping the migration to the generic version.

The generic version also has optional parameters for convenience in setting initial dimensions or default text, when relevant. The default values for all of them are identical to the old method's defaults.

I also moved the two instances of reflection that remained in the ViewFactory out into a separate ReflectionHelpers static class, since they're not really ViewFactory-specific, and potentially helpful for consolidation of other code, as the work to migrate to more ubiquitous generics goes on.

One instance of reflection also got eliminated entirely, as Terminal.Gui fixed the reason it was necessary in the first place (by putting a setter on a property we had to set via reflection, as it had no setter before), so that was a happy finding.

**Reply by dodexahedron** [reply]
The default case, as you discovered, has different behavior for types the original method did not check for.

Fix for that is easy and I'll be sure it gets in there, as I move on to ViewFactory tests.
